### PR TITLE
AuthInfo:Authentication information is updated with the cluster update

### LINKED ISSUE #15510: No password is set for Envoy Listener, but the command is not allowed
**No password is set for Envoy Listener, but the command is not allowed**

@mattklein123 @envoyproxy/redis-dev
 I used go-control-plane to send the configuration to envoy and there was an authorization error.

Envoy starts successfully first and wai for the Pilot to deliver the configuration. After the pilot delivers the configuration, proxy Redis through the envoy。

I test envoyv1.17.0, envoy-1.15.4, the same result under test

image

image

Envoy Listener Config：
```JSON
{
	"name": "listener_33000",
	"active_state": {
		"version_info": "2021-03-16T14:26:15+08:00/1",
		"listener": {
			"@type": "type.googleapis.com/envoy.config.listener.v3.Listener",
			"name": "listener_33000",
			"address": {
				"socket_address": {
					"address": "127.0.0.1",
					"port_value": 33000
				}
			},
			"filter_chains": [{
				"filters": [{
					"name": "envoy.filters.network.redis_proxy",
					"typed_config": {
						"@type": "type.googleapis.com/envoy.extensions.filters.network.redis_proxy.v3.RedisProxy",
						"stat_prefix": "redis_33000",
						"settings": {
							"op_timeout": "5s",
							"enable_redirection": true,
							"enable_command_stats": true
						},
						"prefix_routes": {
							"catch_all_route": {
								"cluster": "33000~Redis"
							}
						}
					}
				}]
			}],
			"traffic_direction": "OUTBOUND"
		},
		"last_updated": "2021-03-16T06:26:15.915Z"
	}
}
```

Envoy Upstream Config：
```json
{
	"version_info": "2021-03-16T14:26:15+08:00/1",
	"cluster": {
		"@type": "type.googleapis.com/envoy.config.cluster.v3.Cluster",
		"name": "33000~Redis",
		"type": "STRICT_DNS",
		"connect_timeout": "1s",
		"lb_policy": "MAGLEV",
		"health_checks": [{
			"timeout": "2s",
			"interval": "2s",
			"unhealthy_threshold": 3,
			"healthy_threshold": 2,
			"custom_health_check": {
				"name": "envoy.health_checkers.redis",
				"typed_config": {
					"@type": "type.googleapis.com/envoy.config.core.v3.HealthCheck.RedisHealthCheck"
				}
			},
			"event_log_path": "/dev/stdout",
			"always_log_health_check_failures": true
		}],
		"circuit_breakers": {
			"thresholds": [{
				"max_connections": 4294967295,
				"max_pending_requests": 4294967295,
				"max_requests": 4294967295,
				"max_retries": 4294967295
			}]
		},
		"dns_refresh_rate": "5s",
		"dns_lookup_family": "V4_ONLY",
		"upstream_connection_options": {
			"tcp_keepalive": {
				"keepalive_probes": 1,
				"keepalive_time": 10,
				"keepalive_interval": 10
			}
		},
		"load_assignment": {
			"cluster_name": "33000~Redis",
			"endpoints": [{
				"lb_endpoints": [{
					"endpoint": {
						"address": {
							"socket_address": {
								"address": "address",
								"port_value": 6379
							}
						}
					}
				}]
			}]
		},
		"typed_extension_protocol_options": {
			"envoy.filters.network.redis_proxy": {
				"@type": "type.googleapis.com/envoy.extensions.filters.network.redis_proxy.v3.RedisProtocolOptions",
				"auth_password": {
					"inline_string": "[redacted]"
				}
			}
		},
		"respect_dns_ttl": true,
		"dns_failure_refresh_rate": {
			"base_interval": "1s",
			"max_interval": "5s"
		}
	},
	"last_updated": "2021-03-16T06:26:15.947Z"
}
```

### PR DESCRIPTION
**Pull request by littlegreyzhang**
Signed-off-by: littlegreyzhang <826999015@qq.com>

Commit Message: AuthInfo:Authentication information is updated with the cluster update

Additional Description:
This changes the authentication information code to update AuthInfo with the cluster update .
Fixes #15510 

Risk Level: Low
Testing: New and existing tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

**Comment by littlegreyzhang** [comment]
@mattklein123 @jmarantz 

Please review this PR . Thank you !

**Comment by littlegreyzhang** [comment]
@envoy-bot ok

**Comment by jmarantz** [comment]
@HenryYYang @msukalski would you be able to take a first pass?

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.cc [hunk 1]
```diff
@@ -130,6 +128,9 @@ void InstanceImpl::ThreadLocalPool::onClusterAddOrUpdateNonVirtual(
 
   ASSERT(cluster_ == nullptr);
   cluster_ = &cluster;
+  // Update username and password when cluster updates
+  auth_username_ = ProtocolOptionsConfigImpl::authUsername(cluster_->info(), parent_.lock()->api_);
+  auth_password_ = ProtocolOptionsConfigImpl::authPassword(cluster_->info(), parent_.lock()->api_);
```
Comment (mattklein123) [inline source/extensions/filters/network/redis_proxy/conn_pool_impl.cc:133]: nit: use shared_parent

File: test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc [hunk 2]
```diff
@@ -523,6 +533,44 @@ TEST_F(RedisConnPoolImplTest, NoClusterAtConstruction) {
   update_callbacks_->onClusterRemoval("fake_cluster");
 }
 
+// ConnPool created when no cluster exists at creation time . Dynamic cluster
+// creation and removal work correctly . Username and password are updated with
+// dynamic cluster .
```
Comment (mattklein123) [inline test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc:538]:
```suggestion
// dynamic cluster.
```

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks this makes sense to me with small comments.

/wait

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.cc (hunk 1 above)
Comment (littlegreyzhang) [reply source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: Got it! Thank you for your suggestion.

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**